codegen/uri.go

```diff
@@ -34,61 +33,77 @@ var (
 
 // ParseModuleURI returns an ast.Module based on the URI provided. The module
 // may live on the local filesystem or remote depending on the scheme.
-func ParseModuleURI(ctx context.Context, cln *client.Client, stdin io.Reader, uri string) (*ast.Module, error) {
-	if uri == "-" {
-		return parser.Parse(ctx, &parser.NamedReader{
-			Reader: stdin,
-			Value:  "<stdin>",
-		})
-	}
-
+func ParseModuleURI(ctx context.Context, cln *client.Client, dir ast.Directory, uri string) (*ast.Module, error) {
 	u, err := url.Parse(uri)
 	if err != nil {
 		return nil, err
 	}
 
 	switch u.Scheme {
 	case "", "file":
-		f, err := os.Open(u.Host + u.Path)
-		if err != nil {
-			return nil, err
-		}
-		defer f.Close()
-		return parser.Parse(ctx, f)
+		return parseModuleFileURI(ctx, cln, dir, u)
 	case "git", "git+https", "git+ssh":
-		return parseModuleGitURI(ctx, cln, u.Scheme, uri)
+		return parseModuleGitURI(ctx, cln, uri)
 	default:
 		return nil, fmt.Errorf("%q is not a valid module uri scheme", u.Scheme)
 	}
 }
 
-func parseModuleGitURI(ctx context.Context, cln *client.Client, scheme, uri string) (*ast.Module, error) {
+func parseModuleFileURI(ctx context.Context, cln *client.Client, dir ast.Directory, u *url.URL) (*ast.Module, error) {
+	filename, err := parser.ResolvePath(ModuleDir(ctx), u.Host+u.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	rc, err := dir.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	mod, err := parser.Parse(ctx, rc)
+	if err != nil {
+		return nil, err
+	}
+	mod.Directory = dir
+
+	if u.Scheme == "" {
+		u.Scheme = "file"
+	}
+	mod.URI = u.String()
+	return mod, nil
+}
+
+func parseModuleGitURI(ctx context.Context, cln *client.Client, uri string) (*ast.Module, error) {
 	u, err := gitscheme.Parse(uri)
 	if err != nil {
 		return nil, err
 	}
+	dockerAPI := DockerAPI(ctx)
+	if dockerAPI.Moby {
+		return nil, errdefs.WithDockerEngineUnsupported(ProgramCounter(ctx))
```

Comment:
We should add docker version (or embedded buildkit version if possible) to `dockerAPI` when `v0.8` finally lands in docker-ce.